### DISCUSSION TITLE
Raster operations

### DISCUSSION THREAD

**Discussion by msbarry** [onthegomap/planetiler#495] [Ideas]
I recently tweaked planetiler to generate an mbtiles of terrarium RGB tiles from https://registry.opendata.aws/terrain-tiles/. I just did this in a one-off branch but I'm curious if people are interested in using planetiler for generating raster tilesets?

A few possibilities I could think of

- copy from one raster tile source (http) to another (mbtiles), applying transformations like reducing the precision, compressing, transforming from png to webp, or merging 256px tiles into 512px one zoom-level lower.
- generate DEM tiles from raw sources (like jaxa, srtm, skadi, etc.)

There are other tools in this space, just curious if there's any appetite for planetiler to do them outside my one-off hacking?

**Comment by wipfli** [comment]
Yeah I would love to use planetiler to create terrain or other raster tiles. I know nothing about raster sources and having a simple way with planetiler to get frontend ready tiles would be great...

**Comment by tyrauber** [comment]
@msbarry, I am very interested in this as well.  I actually need to work up an Outdoor theme.  I've got the NASA ASTER images on an EC2 Volume. I was actually going to run through a series of GDAL contour transformations to produce contour lines, hillshade, etc, and assemble those tiles into a shapefiles, and then convert those to mbiles, etc, and then use them as layers in a Planetiler build profile. But perhaps there is a better way? Is [dem-download](https://github.com/onthegomap/planetiler/compare/main...dem-download) the branch?  Can I help in anyway?

**Reply by msbarry** [reply]
Yes, the [Copier class on dem-download branch](https://github.com/onthegomap/planetiler/blob/dem-download/planetiler-core/src/main/java/com/onthegomap/planetiler/util/Copier.java) is what I was playing with.  It copies from a tile source to tile dest with some transformation.

Some possible goals for raster operations:

1. Generate vector contours + hillshading from DEM sources
    <img width="300" alt="image" src="https://user-images.githubusercontent.com/1480504/224540559-f0931a58-44ac-425a-b81a-f2c2978b4bb6.png">
2. Generate vector landcover from low-zoom satellite or raster landcover source like https://worldcover2021.esa.int/:
    <img width="300" alt="image" src="https://user-images.githubusercontent.com/1480504/224540793-d76b4d40-58a5-4889-ad71-2d271fa34bd6.png">
3. Generate raster tiles from aligning/overlaying raw raster sources (to generate a standard satellite/DEM raster tile output)
4. Transform an existing tiled source (raster DEM -> raster hillshade, raster -> vector, etc.


Here is a high level approach how we could implement these:

![Untitled drawing (5)](https://user-images.githubusercontent.com/1480504/224541538-afd18676-f5d9-4d2d-bfc0-476c50190243.png)


For 1 and 2, we could either

a) use existing gdal tools to map raw raster sources to raw vector sources, then feed those into planetiler. This is more manual work to use, but less upfront work to get to a working version
b) reimplement gdal algorithms inside planetiler so you just point planetiler at raw raster sources and it does the processing/transformation internally. This is more upfront work, but would result in a much easier-to-use end product. For example, planetiler could manage downloading the sources and working with them in-memory so you don't need a ton of extra disk space.

There are also gdal bindings for java that could work as a hybrid between a and b: https://gdal.org/api/java/index.html

**Reply by tyrauber** [reply]
This is fantastic, @msbarry! All those options would be hugely beneficial!

For 1, I think the goal should be processing the highest resolution, freely available GeoTIFF into vector layers - contour lines and hillshading. An optional raster geotiff layer that enables client-side elevation calculation would also be really helpful.

For 2, I would go with a [vector landcover layer](https://worldcover2021.esa.int/). I am not sure what the benefit of a raster landcover layer would be, especially if  the next focus was satellite tiles.

3, Downloading and assembling raster satellite tiles from publicly available sources would be absolutely amazing, and might be of more use to more planetiler users than 1 & 2.

In regards to the integration, it sounds like the java gdal bindings provide the best of both worlds. I think trying to reimplement all the necessary algorithms would be a lot of work on it’s own, and the existing gdal libraries are fast, stable and production ready.  The only reason to implement custom algorithms in my opinion would be if gdal didn’t provide the necessary functionality.

I think one consideration is the data source of the DEM files. [Here is a good overview of the Global Digital Elevation Models](https://eop-cfi.esa.int/index.php/docs-and-mission-data/dem) as of 16 February 2023. 

The terrarium tileset on AWS is great, and freely accessible - a lot of sources require authentication - but I am not sure what the resolution is. It looks like it is compiled from a couple different sources and the output is 51GB.  The ASTER V3 DEM files zipped are 379GB for 1 arcseconds (~30 m). AND it looks like the ASTER V3 DEM files are now freely available!

ASTGTM v003
[ASTER Global Digital Elevation Model 1 arc second](https://lpdaac.usgs.gov/products/astgtmv003/)
[Direct Download](https://e4ftl01.cr.usgs.gov/ASTT/ASTGTM.003/2000.03.01/)

Is there an easy way to translate {z}/{x}/{y} to degrees? A lot of the DEM files come in the following url format:  `ASTGTMV003_N52W177.zip`

**Reply by bdon** [reply]
@tyrauber The DEM would need to be warped into Web Mercator before slicing Z/X/Y.

@msbarry how much code would this share with the rest of Planetiler? I think contour line generation could be interesting as that's inherently vector, but maybe it would make sense for the raster operations to live in a separate codebase?

**Reply by msbarry** [reply]
Thanks @tyrauber - For # 2, I think the goal would be to convert the raster landcover layer from ESA to vector shapes.

I agree gdal bindings seem like a good compromise to get started fast. For contour lines for example, the workflow would be to take a 1°x1° elevation tile, use gdal to generate contour isoline shapes, then pass those through the rest of planetiler's vector shape handling, with some slight tweaks to ensure that contours meet properly at intersections since I think those 1° tiles are slightly overlapping.

That's a great overview of DEM sources. The gzipped 1° "skadi" tiles on AWS are about 200gb total and while that source is 30m, it's built from [sources](https://github.com/tilezen/joerd/blob/master/docs/data-sources.md) that vary in resolution a lot more. [@acalcutt](https://github.com/acalcutt) also pointed me to the JAXA 30m dataset https://www.eorc.jaxa.jp/ALOS/en/aw3d30/data/index.htm although I'm not sure how the quality compares to those others and terrarium?

@bdon here are the categories I see:
- raw raster source->vector tiles processing would share the entire processing pipeline and you may want the layers to end up in the same tileset as layers processed from OSM so definitely makes sense to have that in planetiler
- raw raster source->raster tiles would share tile writing pipeline, and if you are overlaying multiple sources then possibly feature sorting/merging too
- raw raster tile -> raster tile would just share the tile reading/writing code and parallel processing so less of a case to include - except now that we have the unified tile archive writer/readers I was thinking of building some small tools that just copy from readers to writers, optionally applying transformations like merging, stripping vector layers/attributes, and these would be similar to that category.

**Reply by tyrauber** [reply]
Yeah @bdon, we'd need to do a srid translation operation first with either `gdalwarp -s_srs EPSG:4269  -t_srs EPSG:3857` or `gdalwarp -s_srs EPSG:4269  -t_srs EPSG:3857`.   

@msbarry I had the same initial thought as @bdon, regarding whether this operation should be done outside of planetiler and pass the layer in as a shapefile or mbtile?  Couldn't the tool prep and output a file that Planetiler could still use in the processing pipeline, with `addShapefileSource` or something like it?

Each one of these gdal operations is going to require some configuration. 

For example, contours will require at least the interval (-i) and polygon (-p) flag. Hillshading, Slope and Color Relief will require even more configuration including color text files.

Here is some pseudo code of the gdal steps for producing contour lines, and hillshade, slope and color relief.

```
// 3857 projection
gdalwarp -s_srs EPSG:4269 -t_srs EPSG:3857 -r bilinear source.tif source_3857.tif

// contour lines

gdal_contour -a elev source_3857.tif source_contour.shp -i 10.0

// hillshade
gdaldem hillshade -az 45 -z 1.3 source_3857.tif hillshade_3857.tif

// color relief
gdaldem color-relief source_3857.tif color-relief.txt relief_3857.tif

// Slope
gdaldem slope source_3857.tif slope_3857.tif
gdaldem color-relief slope.tif color-slope.txt slope-shade.tif

// Combine
gdalwarp hillshade_3857.tif relief_3857.tif slope_3857.tif output.tif -srcnodata -dstnodata

```

`source_contour.shp` would need to be processed into a vector layer, combining and optimizing the polygons across tiles.  

`output.tif` would need to be translated to a raster layer. As much as I'd love hillshade, slope and color relief as a vector, for client side styling, I just don't know how that would be possible.

References:

- tranform [gdaltransform](https://gdal.org/programs/gdaltransform.html)
- warp [gdalwarp](https://gdal.org/programs/gdalwarp.html)
- contour lines [gdal_contour](https://gdal.org/programs/gdal_contour.html)
- hillshade, relief and slope [gdaldem](https://gdal.org/programs/gdaldem.html)

**Reply by wipfli** [reply]
Landcover and hillshading is nicely combined in a vector layer by google maps for example: 
![Screenshot_20230313-124142_Maps](https://user-images.githubusercontent.com/53421382/224692288-b631c428-83dd-4cc5-bbe2-540ddfc09323.jpg)

**Reply by tyrauber** [reply]
:+1: @wipfli, if we could accomplish that, that would be stellar.
